### ISSUE TITLE
images.jpg file not loading correct bounding box sizes

### ISSUE THREAD

**Issue by michaelglobal** [ultralytics/ultralytics#2383]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

I have created my own dataset to detect sports cards. The bounding box that I created seems to be right in the val_batch1_labels images -
![val_batch1_labels](https://user-images.githubusercontent.com/34948627/235954596-58dd50e2-ab8c-4c0d-bc86-cffd36ff97fc.jpg)

![train_batch0](https://user-images.githubusercontent.com/34948627/235953054-b901bbd4-4830-4c44-8e28-e56812cd2a6a.jpg)

But in the prediction images outputted, the bounding boxes aren't correct.
![val_batch1_pred](https://user-images.githubusercontent.com/34948627/235953121-06da057c-4a88-45e6-8e19-e1f11ac2630a.jpg)

I saw that the labels.jpg saved image, on the right side, is showing the bounding boxes as just squares, which aren't accurate, but is the reason the prediction images are wrong. 
![labels](https://user-images.githubusercontent.com/34948627/235953012-6aefde97-d3e5-4240-8d7e-baa2bc3d24d0.jpg)

Does anybody know why the bounding boxes aren't being correctly logged in the labels.jpg file? Any ideas on how to fix this?

Thank you very much!

### Environment

Python 3.10.9
OS: MacOS Ventura 13.2

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@michaelglobal thanks for your interest in contributing to YOLOv8's development!

Based on the provided images, it appears that the bounding box labels are being saved as squares and, as a result, are not being correctly logged in the labels.jpg file and impacting the accuracy of the generated bounding boxes. 

To further investigate this issue, we recommend that you check that the bounding box labels are being properly saved as rectangles rather than squares in your code. For example, ensure that the coordinates of the top-left and bottom-right corners of each bounding box are properly computed as "x,y" and "x+w,y+h".

If you are able to confirm that this is indeed the issue, you can try to correct the relevant code and submit a PR. We appreciate your support towards the community's effort in maintaining and improving YOLOv8.

Let us know if you have any further questions or need assistance.

**Comment by michaelglobal** [comment]
Here is the code to create the label file:

`label_path = os.path.join(labels_dir, f"{card_file[:-4]}-pos-{i + 1}.txt")
x1, y1 = x_pos, y_pos
x2, y2 = x_pos + w, y_pos + h
x1_norm = (x1 + (w / 2)) / image_size[0]
y1_norm = (y1 + (h / 2)) / image_size[1]
w_norm = (x2 - x1) / image_size[0]
h_norm = (y2 - y1) / image_size[1]
card_type_code = "0" if card_type == "v" else "1"
label = f"{card_type_code} {x1_norm:.6f} {y1_norm:.6f} {w_norm:.6f} {h_norm:.6f}\n"
with open(label_path, "w") as f:
    f.write(label)`

This produces are label like this:
`0 0.281250 0.646793 0.379386 0.394737`

I think the w_norm and h_norm are the  normalized coordinates of the CENTER of the card's bounding box, therefore producing square-like bounding boxes, when I don't use the normalized w and h, the output is this:
`0 0.702577 0.397615 1384.000000 1920.000000`
Which is the card size. These numbers need to be between 0 and 1 but maintain the 'aspect ratio' (the height should be 1.5 time larger than the width). 

Any ideas to output the correct width and height?

Thank you!

**Comment by glenn-jocher** [comment]
Hi @michaelglobal, thank you for your question! 

It looks like your label file is following the correct YOLOv8 format, which requires bounding box labels to include the class and normalized coordinates (`x`, `y`, `width`, `height`) of each object, with one row per object within each image, separated by spaces. 

Regarding your concern with the label file not generating the appropriate normalized bound box size, a bounding box in YOLOv8 is defined by its center coordinates (`x`, `y`) and its width and height, normalized to the dimensions of the original image.

You are correct in your assumption that `w_norm` and `h_norm` in your code denote the normalized coordinates of the center of the card's bounding box, divided by the size of the original image, which results in square-like bounding boxes. To generate the appropriate height and width (while maintaining the aspect ratio) in your label file, we recommend computing the width and height from the bounding box's diagonal, instead of only from the center coordinates. You can compute the diagonal using the Pythagorean theorem, and then compute the rescaled `width` (normalized `width` divided by the diagonal), and `height` (normalized `height` divided by the diagonal multiplied by 1.5 to maintain the aspect ratio).

I hope this helps you tackle your issue! Let us know if you have any further questions.

**Comment by michaelglobal** [comment]
So are the square-like bounding boxes an issue? It's an accurate representation of the bounding boxes of the cards?

Thank you for your replies.

**Comment by glenn-jocher** [comment]
@michaelglobal, thank you for your follow-up question!

Square-like bounding boxes are not an issue if your dataset consists of square objects. However, if your dataset consists of objects with different aspect ratios, such as your case where you need to maintain an aspect ratio of 1:1.5 for your cards, you will need to adjust your `width` and `height` relative to the diagonal of your object's bounding box, instead of only from the center coordinates as previously mentioned.

To recap, square-like bounding boxes in your scenario mean that the dimensions are not proportional to the size of the object, and the accuracy of object detection may be impacted. So you'll want to ensure that you're following YOLOv8's convention of bounding box labels that contain the correct aspect ratio and normalized coordinates.

I hope this answers your question. Let me know if there's anything else you need.
